internal/network/local.go

```diff
@@ -0,0 +1,438 @@
+	return ns.lookup(sock, addr.Addr[:], addr.Port)
+}
+
+func (ns *LocalNamespace) lookup(sock *localSocket, addr net.IP, port int) (*localSocket, error) {
+	for _, iface := range ns.interfaces() {
+		if peer := iface.lookup(sock, addr, port); peer != nil {
+			return peer, nil
+		}
+	}
+
+	if addr.IsUnspecified() {
+		return nil, EHOSTUNREACH
+	}
+
+	n := ns.network.Load()
+	if n == nil {
+		return nil, ENETUNREACH
+	}
+
+	n.mutex.RLock()
+	iface := n.lookup(addr)
+	n.mutex.RUnlock()
+
+	if iface != nil {
+		if peer := iface.lookup(sock, addr, port); peer != nil {
+			return peer, nil
+		}
+		return nil, EHOSTUNREACH
+	}
+	return nil, ENETUNREACH
+}
+
+func (ns *LocalNamespace) unlinkInet4(sock *localSocket, addr *SockaddrInet4) {
+	ns.unlink(sock, addr.Addr[:], addr.Port)
+}
+
+func (ns *LocalNamespace) unlinkInet6(sock *localSocket, addr *SockaddrInet6) {
+	ns.unlink(sock, addr.Addr[:], addr.Port)
+}
+
+func (ns *LocalNamespace) unlink(sock *localSocket, addr net.IP, port int) {
+	for _, iface := range ns.interfaces() {
+		iface.unlink(sock, addr, port)
+	}
+}
+
+type localPort struct {
+	proto Protocol
+	port  uint16
+}
+
+type localInterface struct {
+	index int
+	name  string
+	haddr net.HardwareAddr
+	flags net.Flags
+	addrs []net.IPNet
+
+	mutex sync.RWMutex
+	ports []map[localPort]*localSocket
+}
+
+func (i *localInterface) Index() int {
+	return i.index
+}
+
+func (i *localInterface) MTU() int {
+	return 1500
+}
+
+func (i *localInterface) Name() string {
+	return i.name
+}
+
+func (i *localInterface) HardwareAddr() net.HardwareAddr {
+	return i.haddr
+}
+
+func (i *localInterface) Flags() net.Flags {
+	return i.flags
+}
+
+func (i *localInterface) Addrs() ([]net.Addr, error) {
+	addrs := make([]net.Addr, len(i.addrs))
+	for j := range addrs {
+		addrs[j] = &i.addrs[j]
+	}
+	return addrs, nil
+}
+
+func (i *localInterface) MulticastAddrs() ([]net.Addr, error) {
+	return nil, nil
+}
+
+func (i *localInterface) bind(sock *localSocket, addr net.IP, port int) error {
+	link := localPort{sock.protocol, uint16(port)}
+	ipv4 := addr.To4()
+	name := (Sockaddr)(nil)
+
+	if ipv4 != nil {
+		name = &SockaddrInet4{Addr: ([4]byte)(ipv4), Port: port}
+	} else {
+		name = &SockaddrInet6{Addr: ([16]byte)(addr), Port: port}
+	}
+
+	i.mutex.Lock()
+	defer i.mutex.Unlock()
+
+	if link.port != 0 {
+		for j, a := range i.addrs {
+			if !socketAndInterfaceMatch(addr, a.IP) {
+				continue
+			}
+			if _, used := i.ports[j][link]; used {
+				// TODO:
+				// - SO_REUSEADDR
+				// - SO_REUSEPORT
+				return EADDRNOTAVAIL
+			}
+		}
+	} else {
+		var port int
+	searchFreePort:
+		for port = 49152; port <= 65535; port++ {
+			link.port = uint16(port)
+			for j, a := range i.addrs {
+				if !socketAndInterfaceMatch(addr, a.IP) {
+					continue
+				}
+				if _, used := i.ports[j][link]; used {
+					continue searchFreePort
+				}
+			}
+			break
+		}
+		if port == 65535 {
```

Comment:
```suggestion
		if port == 65536 {
```

should this be `65536` if the loop is exhausted?

Reply:
Correct